proposals/lip-0003.md

```diff
@@ -75,35 +78,32 @@ Given the study of the issue presented in the previous section, this LIP propose
 ## Specification
 
-The changes will only affect the `Delegates.prototype.generateDelegateList` function mentioned above. These changes consist of removing the nested `for` loops and generating a new list of “delegate hashes” to be re-calculated and ordered every round. In particular, the proposed fix implements a new algorithm that goes as follows:
+The changes affect the `shuffleDelegateListForRound` function mentioned above. These changes consist of removing the nested `for` loops and generating a new list of “delegate hashes” to be re-calculated and ordered every round. In particular, the proposed fix is as follows:
 
-1. The variable `seedSource` is generated in the same way as now:
+1. The variable `seedSource` is generated as follows:
 
 	```js
-	var seedSource = modules.rounds.calc(height).toString();
+	const seedSource = randomSeed1;
 	```
+	where `randomSeed1` is the variable specified in [LIP-0022](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0022.md#random-seeds-computation). 
 
-2. One hash is generated per delegate using `seedSource` and the delegate public key (or ID). For simplicity, we propose to use the hashing function already utilised in the existing code, although the final choice here is left to the reference implementation. Example given in pseudo-code:
+2. One hash is generated per delegate using `seedSource` and the delegate address. For simplicity, we propose to use SHA-256, which is the hashing function already utilised in the existing function. Following the current implementation, an example in pseudo-code will be:
 
 	```
-	forEach delegate
-		delegate.pubKeyHash = createHash(seedSource + delegate.publicKey)
+	forEach delegate in delegateList
+		delegate.AddressHash = hash(seedSource + delegate.address)
 	```
+	Note that `+` stands for the string concantenation operator.
 
-3. The delegates are reordered according to the hashes generated in step 2. This can be done according to any criteria (in an ascending order, for example) as long as the list is sorted in a deterministic and consistent way. Hence, the sorting algorithm must be [stable](https://en.wikipedia.org/wiki/Sorting_algorithm#Stability) to generate the same list of delegates on every node on the network, even in the unlikely event of a hash collision in the second point. As an example, the `sort()` method in JavaScript can be used for this purpose.
-
-This will not only fix the uniformity issue (caused by the two issues raised in the [Motivation](#motivation) section), but it will also reduce complexity, allowing simpler, more human-readable code in any implementation.
+3. The delegates are reordered in lexicographical order according to the hashes generated in step 2. This sorting must be deterministic and consistent. This means it has to be [stable](https://en.wikipedia.org/wiki/Sorting_algorithm#Stability) to generate the same list of delegates on every node on the network, even in the unlikely event of a hash collision in the second point.
```

Comment:
We would need to define the order of the delegate list before calling the sorting *stable* algorithm so that tie-breaking in case of a hash collision is defined. Originally, the delegate list was always given by rank. With standby delegates, we may want to assume lexicographic order by address before the sorting.

Reply:
but then wouldn't we be assuming no collision in the address (which would imply collision in the hash if I understand correctly)? So that would not solve the problem. Thinking about it, the problem is not hash collision of 32 bytes, but of 20 bytes (maybe this was clear)

Reply:
disregard my previous comments, addresses are unique identifiers

Reply:
I think it is an overkill to impose a certain order in the input or to do an extra sorting before the actual ordering for this reason. Here we would need to have a hash collision AND a different input list order in the nodes (in practice this input list should be the same for every node if they run our software). I think we are safe by just assuming that SHA-256 is collision resistant (as it is done normally).  Therefore I am going to remove the last sentences of this bullet point, that by the way they are more rationale than specification.